src/mutex.hpp

```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
+  // exclusive lock.
+  void lock() ACQUIRE(){std_mutex.lock();}
+  // Try to acquire the mutex.  Returns true on success, and false on failure.
+  bool try_lock() TRY_ACQUIRE(true){return std_mutex.try_lock();}
+
+  // Acquire/lock this mutex for read operations, which require only a shared
+  // lock.  This assumes a multiple-reader, single writer semantics.  Multiple
+  // threads may acquire the mutex simultaneously as readers, but a writer
+  // must wait for all of them to release the mutex before it can acquire it
+  // exclusively.
+  void lock_shared() ACQUIRE_SHARED(){std_mutex.lock_shared();}
+  // Try to acquire the mutex for read operations.
+  bool try_lock_shared() TRY_ACQUIRE_SHARED(true) { return std_mutex.try_lock_shared(); }
+
+  // Release/unlock an exclusive mutex.
+  void unlock() RELEASE(){std_mutex.unlock();}
+
+  // Release/unlock a shared mutex.
+  void unlock_shared() RELEASE_SHARED(){std_mutex.unlock_shared();}
+
+  // Assert that this mutex is currently held by the calling thread.
+//  void AssertHeld() ASSERT_CAPABILITY(this);
+
+  // Assert that is mutex is currently held for read operations.
+//  void AssertReaderHeld() ASSERT_SHARED_CAPABILITY(this);
+
+  // For negative capabilities.
+//  const Mutex& operator!() const { return *this; }
+};
+
+
 // MutexLocker is an RAII class that acquires a mutex in its constructor, and
 // releases it in its destructor.
 template <class t_mutex>
-class SCOPED_CAPABILITY Lock_guard {
+class SCOPED_CAPABILITY LockGuard {
 private:
   t_mutex &mut;
 
 public:
-  Lock_guard(t_mutex &mu) ACQUIRE(mu) : mut(mu) {
-    mut.lock();
-  }
-  ~Lock_guard() RELEASE() {
-    mut.unlock();
-  }
+  LockGuard(t_mutex &mu) ACQUIRE(mu) : mut(mu) { mut.lock(); }
+  ~LockGuard() RELEASE() { mut.unlock(); }
+  void lock() ACQUIRE() { mut.lock(); }
+  void unlock() RELEASE() { mut.unlock();  }
 };
 
 template <class t_mutex>
-class SCOPED_CAPABILITY Unique_lock {
+class SCOPED_CAPABILITY UniqueLockGuardRO {
 private:
-  std::unique_lock<t_mutex> unique_lock;
-
+  t_mutex & m_mut;
+  bool m_locked; ///< not all mutex types know if they are locked apparentyl?
 public:
-  Unique_lock(t_mutex &mu) ACQUIRE(mu) : unique_lock(mu) {}
+  UniqueLockGuardRO(t_mutex &mu) noexcept ACQUIRE_SHARED() : m_mut(mu), m_locked(true) {
+  	try { m_mut.lock_shared(); } catch(...) { _mutexshield_abort("LG-RO constr"); } }
+  ~UniqueLockGuardRO() noexcept RELEASE_SHARED() {
+  	try { if (m_locked) m_mut.unlock_shared(); } catch(...) { _mutexshield_abort("LG-RO destr"); } }
+
+  void lock() noexcept RELEASE_SHARED() {
+  	try {
+  		if ( m_locked) _mutexshield_abort("LG-RO already locked");
+	  	m_mut.unlock_shared(); m_locked=false;
+	  } catch(...) { _mutexshield_abort("LG-RO lock"); } }
+  void unlock() noexcept RELEASE_SHARED() {
+  	try {
+  		if (!m_locked) _mutexshield_abort("LG-RO already unlocked");
+	  	m_mut.unlock_shared(); m_locked=false;
```

Comment:
NACK, (disc), it's unique lock (not a mutex) so it's ok here

no one can call two .unlock() on this unique lock
this classes are not thread-safe